File: app/src/main/java/com/intprog/helpinghands/screens/PaymentPageActivity.java
package com.intprog.helpinghands.screens;

import androidx.appcompat.app.AppCompatActivity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.Toast;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.intprog.helpinghands.HomePageActivity;
import com.intprog.helpinghands.ProfilePageActivity;
import com.intprog.helpinghands.R;
import com.intprog.helpinghands.screens.DonationCampaign.DonationOptionPageActivity;
import com.stripe.android.PaymentConfiguration;
import com.stripe.android.paymentsheet.PaymentSheet;
import com.stripe.android.paymentsheet.PaymentSheetResult;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.HashMap;
import java.util.Map;

public class PaymentPageActivity extends AppCompatActivity {

    PaymentSheet paymentSheet;
    String paymentIntentClientSecret;
    PaymentSheet.CustomerConfiguration customerConfig;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_payment_page);

        // ImageButton for navigating to DonationOptionPageActivity
        ImageButton backTop = findViewById(R.id.backTop);
        backTop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(PaymentPageActivity.this, DonationOptionPageActivity.class);
                startActivity(intent);
                finish();
            }
        });

        // ImageButton for navigating to HomePageActivity
        ImageButton homeImageButton = findViewById(R.id.homeImageButton);
        homeImageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(PaymentPageActivity.this, HomePageActivity.class);
                startActivity(intent);
            }
        });

        // ImageButton for navigating to ProfilePageActivity
        ImageButton profileImageButton = findViewById(R.id.profileImageButton);
        profileImageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(PaymentPageActivity.this, ProfilePageActivity.class);
                startActivity(intent);
            }
        });

        fetchPaymentApi();

        Button button = findViewById(R.id.btn_pay);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (paymentIntentClientSecret != null)
                    paymentSheet.presentWithPaymentIntent(paymentIntentClientSecret, new PaymentSheet.Configuration("Codes Easy", customerConfig));
                else
                    Toast.makeText(getApplicationContext(), "API Loading..", Toast.LENGTH_SHORT).show();
            }
        });

        paymentSheet = new PaymentSheet(this, this::onPaymentSheetResult);
    }

    public void fetchPaymentApi() {
        RequestQueue queue = Volley.newRequestQueue(getApplicationContext());
        String url = "https://helpinghands-stripe-api.onrender.com/";

        StringRequest stringRequest = new StringRequest(Request.Method.POST, url,
                new Response.Listener<String>() {
                    @Override
                    public void onResponse(String response) {
                        Log.e("R", response);
                        try {
                            final JSONObject result = new JSONObject(response);
                            customerConfig = new PaymentSheet.CustomerConfiguration(
                                    result.getString("customer"),
                                    result.getString("ephemeralKey")
                            );
                            paymentIntentClientSecret = result.getString("paymentIntent");
                            PaymentConfiguration.init(getApplicationContext(), result.getString("publishableKey"));

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
            }
        }) {
            @Override
            protected Map<String, String> getParams() {
                Map<String, String> paramV = new HashMap<>();
                paramV.put("authKey", "abc");
                return paramV;
            }
        };
        queue.add(stringRequest);
    }

    private void onPaymentSheetResult(final PaymentSheetResult paymentSheetResult) {
        if (paymentSheetResult instanceof PaymentSheetResult.Canceled) {
            Log.e("R:", "Canceled");
        } else if (paymentSheetResult instanceof PaymentSheetResult.Failed) {
            Log.e("App", "Got error: ", ((PaymentSheetResult.Failed) paymentSheetResult).getError());
        } else if (paymentSheetResult instanceof PaymentSheetResult.Completed) {
            // Display for example, an order confirmation screen
            Log.e("R:", "Completed");
        }
    }
}

File: app/src/main/java/com/intprog/helpinghands/screens/PaymentPageActivity.kt
package com.intprog.helpinghands.screens

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.github.kittinunf.fuel.Fuel
import com.github.kittinunf.fuel.core.FuelError
import com.github.kittinunf.fuel.core.Handler
import com.intprog.helpinghands.HomePageActivity
import com.intprog.helpinghands.ProfilePageActivity
import com.intprog.helpinghands.R
import com.intprog.helpinghands.screens.DonationCampaign.DonationOptionPageActivity
import com.stripe.android.PaymentConfiguration
import com.stripe.android.paymentsheet.PaymentSheet
import com.stripe.android.paymentsheet.PaymentSheetResult
import org.json.JSONException
import org.json.JSONObject

class PaymentPageActivity : AppCompatActivity() {

    private lateinit var stripeButton: Button
    private lateinit var amountEditText: EditText
    private lateinit var backTop: ImageButton
    private lateinit var homeImageButton: ImageButton
    private lateinit var profileImageButton: ImageButton

    private lateinit var paymentSheet: PaymentSheet
    private lateinit var paymentIntentClientSecret: String
    private lateinit var amount: String
    private lateinit var customerConfig: PaymentSheet.CustomerConfiguration

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_payment_page)

        // Initialize buttons
        stripeButton = findViewById(R.id.stripeButton)
        amountEditText = findViewById(R.id.amountEditText)
        backTop = findViewById(R.id.backTop)
        homeImageButton = findViewById(R.id.homeImageButton)
        profileImageButton = findViewById(R.id.profileImageButton)

        // Set click listeners for navigation buttons
        backTop.setOnClickListener {
            val intent = Intent(this@PaymentPageActivity, DonationOptionPageActivity::class.java)
            startActivity(intent)
            finish()
            overridePendingTransition(0, 0)
        }

        homeImageButton.setOnClickListener {
            val intent = Intent(this@PaymentPageActivity, HomePageActivity::class.java)
            startActivity(intent)
            overridePendingTransition(0, 0)
        }

        profileImageButton.setOnClickListener {
            val intent = Intent(this@PaymentPageActivity, ProfilePageActivity::class.java)
            startActivity(intent)
            overridePendingTransition(0, 0)
        }

        stripeButton.setOnClickListener {
            if (amountEditText.text.toString().isEmpty()) {
                Toast.makeText(this@PaymentPageActivity, "Amount cannot be empty", Toast.LENGTH_SHORT).show()
            } else {
                val amountStr = amountEditText.text.toString()
                if (isAmountValid(amountStr)) {
                    amount = amountStr + "00"
                    getDetails()
                }
            }
        }

        paymentSheet = PaymentSheet(this@PaymentPageActivity, ::onPaymentSheetResult)
    }

    private fun isAmountValid(amountStr: String): Boolean {
        val amountValue = amountStr.toInt()
        if (amountValue < 100) {
            Toast.makeText(this@PaymentPageActivity, "Amount cannot be below 100", Toast.LENGTH_SHORT).show()
            return false
        }
        return true
    }

    private fun getDetails() {
        Fuel.post("https://us-central1-helpinghands-96b47.cloudfunctions.net/HelpingHands?amt=$amount")
            .responseString(object : Handler<String> {
                override fun success(s: String) {
                    try {
                        val result = JSONObject(s)
                        customerConfig = PaymentSheet.CustomerConfiguration(
                            result.getString("customer"),
                            result.getString("ephemeralKey")
                        )
                        paymentIntentClientSecret = result.getString("paymentIntent")
                        PaymentConfiguration.init(applicationContext, result.getString("publishableKey"))

                        runOnUiThread { showStripePaymentSheet() }

                    } catch (e: JSONException) {
                        Toast.makeText(
                            this@PaymentPageActivity,
                            e.message,
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }

                override fun failure(fuelError: FuelError) {
                    // Handle failure
                }
            })
    }

    private fun showStripePaymentSheet() {
        val configuration = PaymentSheet.Configuration.Builder("HelpingHands")
            .customer(customerConfig)
            .allowsDelayedPaymentMethods(true)
            .build()
        paymentSheet.presentWithPaymentIntent(
            paymentIntentClientSecret,
            configuration
        )
    }

    private fun onPaymentSheetResult(paymentSheetResult: PaymentSheetResult) {
        when (paymentSheetResult) {
            is PaymentSheetResult.Canceled -> Toast.makeText(
                this@PaymentPageActivity,
                "Payment Cancelled",
                Toast.LENGTH_SHORT
            ).show()
            is PaymentSheetResult.Failed -> Toast.makeText(
                this@PaymentPageActivity,
                (paymentSheetResult as PaymentSheetResult.Failed).error.toString(),
                Toast.LENGTH_SHORT
            ).show()
            is PaymentSheetResult.Completed -> Toast.makeText(
                this@PaymentPageActivity,
                "Payment Successful",
                Toast.LENGTH_SHORT
            ).show()
        }
    }
}
